scheduler/test/cook/test/mesos/mesos_mock.clj

```diff
@@ -0,0 +1,844 @@
+                 "end_time_ms" (.getTime (or (:instance/end-time task) (java.util.Date.)))
+                 "status" (:instance/status task)
+                 "hostname" (:instance/hostname task)
+                 "slave_id" (:instance/slave-id task)})]
+    (with-open [out-file (io/writer file)]
+      (csv/write-csv out-file
+                     (concat [headers]
+                             (map #(vals (select-keys % headers)) tasks))))))
+
+(deftest cook-scheduler-integration
+  ;; This tests the the case where one user has all the resources in the cluster
+  ;; and then another user shows up which causes the scheduler to preempt jobs
+  ;; so both may run.
+  ;; TODO: Explicitly check that both users have ~same resources instead of
+  ;;       simply checking that the first user had some jobs preempted
+  (testing "basic scheduling / rebalancing test"
+    (log/info "starting cook scheduler integration test")
+    (let [mesos-datomic-conn (restore-fresh-database! (str (gensym "datomic:mem://mock-mesos")))
+          mem 2000.0
+          cpus 2.0
+          ports [{:begin 1 :end 100}]
+          num-hosts 10
+          hosts (for [_ (range num-hosts)]
+                  (make-random-host :mem {"*" mem} :cpus {"*" cpus} :ports {"*" ports}))
+          speed-multiplier 20
+          make-mesos-driver-fn (fn [scheduler _];; _ is framework-id
+                                 (mm/mesos-driver-mock hosts speed-multiplier scheduler))]
+      (with-cook-scheduler mesos-datomic-conn make-mesos-driver-fn {}
+        (share/set-share! mesos-datomic-conn "default" :mem mem :cpus cpus :gpus 1.0)
+        ;; Note these two vars are lazy, need to realize to put them in db.
+        (let [user-a-job-ent-ids (for [_ (range 20)]
+                                   (create-dummy-job mesos-datomic-conn
+                                                     :user "a"
+                                                     :command "10000"
+                                                     :custom-executor? false
+                                                     :memory mem
+                                                     :ncpus cpus))
+
+              user-b-job-ent-ids (for [_ (range 5)]
+                                   (create-dummy-job mesos-datomic-conn
+                                                     :user "b"
+                                                     :command "1000"
+                                                     :custom-executor? false
+                                                     :memory mem
+                                                     :ncpus cpus))
+              update-status-error-ms 200
+              ]
+          ;; Transact user "a"s jobs
+          (doall user-a-job-ent-ids)
+          ;; Let them get scheduled
+          (Thread/sleep 5000)
+          ;; Transact user "b"s jobs
+          (doall user-b-job-ent-ids)
+          ;; Let the system complete
+          (Thread/sleep 30000)
```

Comment:
This one is a little trickier to remove because of the polling that would need to be done. I used the tracing csv to make close bounds so it shouldn't be too much slower. In tests, I was seeing things complete in ~22-25 seconds and so gave a little bit of a buffer. Can we push this off for a subsequent PR?